File: Python/diffusion.py
#!/usr/bin/python3
#imported libraries
import sys
import math
from numpy import *



#main function for all code 
def main() :
  #intialize maximum cube size
  maxsize = int(input("Enter the maximum size: "))
  
  #zeroing the cube with a loop
  cube = zeros((maxsize, maxsize, maxsize), dtype = 'double')   # Initialize a 3D-array with zeroes
  
  
  #adding partition
  flag = input("Would you like to add a partition to this room (y / n) : ")
  if (flag == "y") :
    flag = True
  elif (flag == "n") :
    flag = False
  if (flag) :
    px = (math.ceil(maxsize * .5)) - 1
    py = (math.ceil(maxsize * .25)) - 1    
    for i in range (py, maxsize) :
      for j in range (0, maxsize) :
        cube[px][i][j] = -1.0
        
  
  #intializing variables
  diffusion_coefficient = 0.175
  room_dimension = 5
  speed_of_gas_molecules = 250.0
  timestep = (room_dimension / speed_of_gas_molecules) / maxsize
  distance_between_blocks = room_dimension / maxsize
  DTerm = diffusion_coefficient * timestep / (distance_between_blocks*distance_between_blocks)
  time = 0.0
  ratio = 0.0
  
  #initialize first cell
  cube[0][0][0] = 1.0e21
  
  #Main loop 
  while ratio < 0.99 :          
    for i in range(0, maxsize) :
      for j in range(0, maxsize) :
        for k in range(0, maxsize) :
          for l in range(0, maxsize) :
            for m in range(0, maxsize) :
              for n in range(0, maxsize) :
                if (((i == l) and (j == m) and (k == n+1) or
                     (i == l) and (j == m) and (k == n-1) or 
                     (i == l) and (j == m+1) and (k == n) or
                     (i == l) and (j == m-1) and (k == n) or 
                     (i == l+1) and (j == m) and (k == n) or
                     (i == l-1) and (j == m) and (k == n))) :
                       if (cube[i][j][k] != -1 and cube[l][m][n] != -1) :   ##Check if there is wall
                         change = (cube[i][j][k] - cube[l][m][n]) * DTerm
                         cube[i][j][k] = cube[i][j][k] - change
                         cube[l][m][n] = cube[l][m][n] + change
    
    ##adjusting time variable                     
    time = time + timestep
    
    ##finding min / max values
    sumval = 0.0
    maxval = cube[0][0][0]
    minval = cube[0][0][0]
    for i in range(0, maxsize) :
      for j in range(0, maxsize) :
        for k in range(0, maxsize) : 
          if (cube[i][j][k] != -1) :          ##Check if there is wall
            maxval = max(cube[i][j][k],maxval)
            minval = min(cube[i][j][k],minval)
            sumval += cube[i][j][k]
    
    #Calculating the ratio      
    ratio = minval / maxval
    
    #Printing the ratio
    print( ratio , " time = " , time )
    #print( time , " " , cube[0][0][0])
    #print(" " , cube[maxsize-1][0][0]) 
    #print(" " , cube[maxsize-1][maxsize-1][0])
    #print(" " ,cube[maxsize-1][maxsize-1][maxsize-1])
    #print(" " , sumval , "\n")
    
    #printing last statement
  print( "Box equilibrated in " + str(time) + " seconds of simulated time." )
  
#call to main function
main()

File: Julia/diffusion.jl
#!/usr2/local/julia-1.8.2/bin/julia


#Ask for maximum size
print("What is the dimension of your room: ")
dimension = readline()
dimension = parse(Int64, dimension)

#Declaring cube and set it to zero
cube = zeros(dimension, dimension, dimension)


#Adding Partition
print("Would you like to add a partition to the room (y/n) : ")
  x = readline()
  if (x == "y")
   return 1
  else 
   return 0 
  end
flag = x

# if flag is equal to y(1)
if (flag == 1)
  px = ceil(Int64, dimension * .5)
  py = ceil(Int64, dimension * .25) 
  for j = py:dimension
    for k = 1:dimension 
      cube[px, j, k] = -1.0       #adding wall of partition of -1
    end
  end  
end


#Declaring variables
diffusion_coefficient = 0.175
room_dimension = 5
speed_of_gas_molecules = 250.0
timestep = (room_dimension / speed_of_gas_molecules) / dimension
distance_between_blocks = room_dimension / dimension
d_term = diffusion_coefficient * timestep / (distance_between_blocks * distance_between_blocks)

#Et cube's first value
cube[1, 1, 1] = 1.0e21

time = 0.0
ratio = 0.0

#Main Loop
while ratio < 0.99

  for i = 1:dimension
    for j = 1:dimension
      for k = 1:dimension
        for l = 1:dimension
          for m = 1:dimension
            for n = 1:dimension
              if i == l && j == m && k == n + 1 ||
                 i == l && j == m && k == n - 1 ||
                 i == l && j == m + 1 && k == n ||
                 i == l && j == m - 1 && k == n ||
                 i == l + 1 && j == m && k == n ||
                 i == l - 1 && j == m && k == n 
                 if cube[i, j, k] != -1 && cube[l, m, n] != -1            #Check for wall
                     global change = ( cube[i, j, k] - cube[l, m, n] ) * d_term
                     global cube[i, j, k] = cube[i, j, k] - change
                     global cube[l, m, n] = cube[l, m, n] + change
                 end
              end
            end
          end
        end
      end
    end
  end

  #updates global time variable
  global time = time + timestep
  
  sumval = 0.0
  
  #intializing min and max values
  maxval = cube[1, 1, 1]  
  minval = cube[1, 1, 1]
  for i = 1:dimension
    for j = 1:dimension
      for k = 1:dimension
          if cube[k,j,i] == -1                  # check for wall
              continue
          end
          maxval = max(cube[i, j, k], maxval)
          minval = min(cube[i, j, k], minval)
          sumval = sumval + cube[i, j, k]
      end
    end 
  end 
  
  #updating global ratio variable
  global ratio = minval / maxval
  
  #priniting the ratio and time for the user
  print(ratio)
  print("        ")
  println(time)

end  


  #printing end statement
  println("The box equilibrated in $time seconds of simulation time.")

exit()



